flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/SimplifyWindowTableFunctionRule.java

```diff
@@ -0,0 +1,355 @@
+                leftWindowTVFRel = rels.get(3);
+            } else {
+                throw new TableException("This should never happen. Please file an issue.");
+            }
+            final StreamPhysicalWindowTableFunction leftWindowTVF =
+                    (StreamPhysicalWindowTableFunction) leftWindowTVFRel;
+            final StreamPhysicalWindowTableFunction rightWindowTVF =
+                    (StreamPhysicalWindowTableFunction) rels.get(rels.size() - 1);
+            return needSimplify(leftWindowTVF) || needSimplify(rightWindowTVF);
+        } else {
+            throw new IllegalStateException(
+                    this.getClass().getName()
+                            + " matches a wrong relation tree: "
+                            + RelOptUtil.toString(node));
+        }
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        List<RelNode> rels = call.getRelList();
+        final RelNode node = rels.get(0);
+        if (node instanceof StreamPhysicalWindowRank) {
+            RelNode newTree = rebuild(rels);
+            call.transformTo(newTree);
+        } else if (node instanceof StreamPhysicalWindowJoin) {
+            RelNode newLeft;
+            RelNode newRight;
+            if (call.rule == WITH_WINDOW_JOIN) {
+                newLeft = rebuild(rels.subList(1, 3));
+                newRight = rebuild(rels.subList(3, 5));
+            } else if (call.rule == WITH_RIGHT_CALC_WINDOW_JOIN) {
+                newLeft = rebuild(rels.subList(1, 3));
+                newRight = rebuild(rels.subList(3, 6));
+            } else if (call.rule == WITH_LEFT_CALC_WINDOW_JOIN) {
+                newLeft = rebuild(rels.subList(1, 4));
+                newRight = rebuild(rels.subList(4, 6));
+            } else if (call.rule == WITH_LEFT_RIGHT_CALC_WINDOW_JOIN) {
+                newLeft = rebuild(rels.subList(1, 4));
+                newRight = rebuild(rels.subList(4, 7));
+            } else {
+                throw new TableException("This should never happen. Please file an issue.");
+            }
+            final StreamPhysicalWindowJoin join = call.rel(0);
+            final RelNode newJoin =
+                    join.copy(
+                            join.getTraitSet(),
+                            join.getCondition(),
+                            newLeft,
+                            newRight,
+                            join.getJoinType(),
+                            join.isSemiJoinDone());
+            call.transformTo(newJoin);
+        } else {
+            throw new IllegalStateException(
+                    this.getClass().getName()
+                            + " matches a wrong relation tree: "
+                            + RelOptUtil.toString(node));
+        }
+    }
+
+    private boolean needSimplify(StreamPhysicalWindowTableFunction windowTVF) {
+        return windowTVF.windowing().isRowtime() && !windowTVF.emitPerRecord();
```

Comment:
```suggestion
        return windowTVF.windowing().isRowtime() && windowTVF.emitPerRecord();
```
Do I understand right?👆

Reply:
using !windowTVF.emitPerRecord() to exclude those window table function nodes which are already emit per record.  Because there is no need to create new nodes for those window table function nodes.
I could add some comments here to explain logic.

Reply:
Thanks for your explanation. I got your meaning.